File: countseeds.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
###############################################################################################################
#
#  Filippos Klironomos, Department of Pediatric Hematology, Oncology and SCT Charité University Hospital Berlin
#
###############################################################################################################
usage = """
    this script takes FASTA sequences of miRNA seeds, groups them by miRNA name, and counts the number of perfect
    matches on the corresponding target FASTA sequences.
    All matches of seeds of the same miRNA group that overlap by at least one nucleotide are counted as single matches.
    
    Output is in TSV format:

        (miRNA name)\t(target name)\t(non-zero counts)
"""
###################################################################################################

import sys,re
from Bio import SeqIO
from Bio.Seq import Seq
from collections import defaultdict
from itertools import izip
from Rfunctions import unlist
import seqMotifs as sm
from argparse import *

parser = ArgumentParser(usage=usage)
parser.add_argument('--targets', default='targets.fa', action='store', dest='targets', help='FASTA file of the named target sequences')
parser.add_argument('--seeds', default='seeds.fa', action='store', dest='seeds', help='FASTA file of the named seed sequences')
parser.add_argument('output', default='counts.tsv', action='store', help='output file where non-zero counts are stored')
options = parser.parse_args()


#  read target sequences
print 'loading the target sequences...'
f=open(options.targets, 'r')
targets=[]
_=[targets.extend([s.id, str(s.seq)]) for s in SeqIO.parse(f, 'fasta')]
f.close()
print 'done!\n'


#  read seed sequences
print 'loading the seed sequences...'
f=open(options.seeds, 'r')
seeds=[]
_=[seeds.extend([s.id, str(s.seq)]) for s in SeqIO.parse(f, 'fasta')]
f.close()
print 'done!\n'


print 'starting the matching...'
f=open(options.output, 'w')


#  capitalize sequences and replace U with T if applicable
seeds[1::2] = [ s.upper().replace('U','T') for s in seeds[1::2] ]
targets[1::2] = [ s.upper().replace('U','T') for s in targets[1::2] ]


#import ipdb;  ipdb.set_trace()
#from IPython.core.debugger import Pdb; ipdb=Pdb(); ipdb.set_trace()


#  make a dictionary with keys the miRNA names and values the corresponding seeds
names=defaultdict(list)
for n,s in izip( seeds[0::2], seeds[1::2] ):
    names[n].append(s)


#  go over all seeds of each miRNA and match them against all targets
for (seeds_name,all_seeds) in names.items():
    
    #  compile seeds as regular expressions (case-sensitive since we have capitalized everything)
    regs=[re.compile(s) for s in all_seeds]  
    

    #  go over all targets and find matches for the given seeds
    for (target_name, target_seq) in izip( targets[0::2], targets[1::2] ):
        
        #  find possible matches
        matches = unlist([sm.myfinditer(r,target_seq) for r in regs]) 
    
        
        #  merge overlapping ranges if we have matches
        if len(matches)!=0:
            matches = sm.reduce_ranges(matches)


            #  write the entry
            f.writelines('\t'.join([seeds_name, target_name, str(len(matches))])+'\n')


f.close()
print 'done!\n'

File: countseeds.R
###############################################################################################################
#
#  Filippos Klironomos, Department of Pediatric Hematology, Oncology and SCT Charité University Hospital Berlin
#
###############################################################################################################




countseeds<-function(SEEDS, SEQS){
    #  this function counts the occurrence of miRNA seeds (SEEDS) in sequences (SEQS) even when the seeds of a 
    #  given miRNA match overlapping positions.
    #
    #  !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
    #  !                                                                              !
    #  !  Generally one would want to count overlapping seed positions as one match,  !
    #  !  using the following python script:                                          !
    #  !                                                                              !
    #  !      countseeds.py --targets targets.fa --seeds seeds.fa counts.tsv          !
    #  !                                                                              !
    #  !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
    #
    #  SEEDS = DNAStringSet of identical length miRNA seeds (reverse-complemented) named after the corresponding miRNA 
    #          different seeds with same miRNA name are allowed
    #  SEQS = DNAStringSet of named sequences where seed occurrences will be counted on
    require(data.table)
    require(Biostrings)


    #  convert to strings since rownames will be defined by them and matchPattern() functions do not allow for duplicate names
    M<-data.table(data.frame(seed=unname(sapply(SEEDS, toString)), mir=names(SEEDS)))


    #  count seed occurrences
    M.pd<-PDict(M$seed)
    v<-t(as.data.frame(vcountPDict(M.pd, SEQS, max.mismatch=0, with.indels=F, fixed=T)))
    rownames(v)<-names(SEQS)
    colnames(v)<-M$seed


    #  collect transcript names and seed counts per seed 
    v<-apply(v, 2, function(x){ y<-x[x!=0];  data.table(tx_name=list(names(y)), ncount=list(unname(y))) } )
    stopifnot( all.equal( names(v), M$seed ) )
    v<-do.call(rbind, v)
    v$seed<-M$seed
    v$mir<-M$mir


    #  uncollapse transcript names and counts to have unique entries for each transcript:seed interaction
    #  N.B. this eliminates empty interactions without a transcript or count
    v<-v[, .(tx_name=unlist(tx_name), ncount=unlist(ncount)), by=.(seed, mir)]


    #  sum over seed counts for given miRNA:transcript interaction
    v<-v[, .(seed=list(seed), ncount=sum(ncount)), by=.(mir, tx_name)]


    #  reorder the columns
    v<-v[, c('mir', 'tx_name', 'seed', 'ncount'), with=F]


    return(v)
}

